x-pack/plugins/monitoring/server/telemetry_collection/get_logstash_stats.ts

```diff
@@ -0,0 +1,423 @@
+  { page = 0, ...options }: { page?: number } & LogstashProcessOptions
+): Promise<void> {
+  const params = {
+    headers: {
+      'X-QUERY-SOURCE': TELEMETRY_QUERY_SOURCE,
+    },
+    index: INDEX_PATTERN_LOGSTASH,
+    ignoreUnavailable: true,
+    filterPath: [
+      'hits.hits._source.logstash_state.pipeline.batch_size',
+      'hits.hits._source.logstash_state.pipeline.workers',
+      'hits.hits._source.logstash_state.pipeline.representation.graph.vertices.config_name',
+      'hits.hits._source.logstash_state.pipeline.representation.graph.vertices.plugin_type',
+      'hits.hits._source.logstash_state.pipeline.representation.graph.vertices.meta.source.protocol',
+      'hits.hits._source.logstash_state.pipeline.representation.graph.vertices',
+      'hits.hits._source.type',
+    ],
+    body: {
+      query: createQuery({
+        filters: [
+          { terms: { 'logstash_state.pipeline.ephemeral_id': ephemeralIds } },
+          {
+            bool: {
+              must: { term: { type: 'logstash_state' } },
+            },
+          },
+        ],
+      }),
+      from: page * HITS_SIZE,
+      collapse: { field: 'logstash_state.pipeline.ephemeral_id' },
+      sort: [{ ['timestamp']: { order: 'desc', unmapped_type: 'long' } }],
+      size: HITS_SIZE,
+    },
+  };
+
+  const results = await callCluster<SearchResponse<LogstashState>>('search', params);
+  const hitsLength = results?.hits?.hits.length || 0;
+  if (hitsLength > 0) {
+    // further augment the clusters object with more stats
+    processLogstashStateResults(results, clusterUuid, options);
+
+    if (hitsLength === HITS_SIZE) {
+      // call recursively
+      const nextOptions = {
+        page: page + 1,
+        ...options,
+      };
+
+      // returns a promise and keeps the caller blocked from returning until the entire clusters object is built
+      return fetchLogstashState(callCluster, clusterUuid, ephemeralIds, nextOptions);
+    }
+  }
+  return Promise.resolve();
+}
+
+export interface LogstashStatsByClusterUuid {
+  [clusterUuid: string]: LogstashBaseStats;
+}
+
+/*
+ * Call the function for fetching and summarizing Logstash stats
+ * @return {Object} - Logstash stats in an object keyed by the cluster UUIDs
+ */
+export async function getLogstashStats(
+  callCluster: LegacyAPICaller,
+  clusterUuids: string[]
+): Promise<LogstashStatsByClusterUuid> {
+  const options: LogstashProcessOptions = {
+    clusters: {}, // the result object to be built up
+    allEphemeralIds: {},
+    versions: {},
+    plugins: {},
+  };
+
+  await Promise.all([fetchLogstashStats(callCluster, clusterUuids, options)]).then(() => {
+    Promise.all([
+      clusterUuids.forEach((clusterUuid) => {
+        if (options.clusters[clusterUuid] !== undefined) {
+          fetchLogstashState(
+            callCluster,
+            clusterUuid,
+            options.allEphemeralIds[clusterUuid],
+            options
+          );
+        }
+      }),
+    ]);
+  });
```

Comment:
No need for the `.then` chain. `await` already ensures that one promise succeeds before the next one. Also, we need to return the array of promises so we can await them. Otherwise, the function will return before waiting for them to be fulfilled.

```suggestion
  await fetchLogstashStats(callCluster, clusterUuids, options);
  await Promise.all(
    clusterUuids.map((clusterUuid) => {
      if (options.clusters[clusterUuid] !== undefined) {
        return fetchLogstashState(
          callCluster,
          clusterUuid,
          options.allEphemeralIds[clusterUuid],
          options
        );
      }
    })
  );
```